src/osd/mClockPoolQueue.h

```diff
@@ -0,0 +1,116 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SK Telecom
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+
+#pragma once
+
+#include <ostream>
+
+#include "boost/variant.hpp"
+
+#include "common/config.h"
+#include "common/ceph_context.h"
+#include "common/mClockPriorityQueue.h"
+#include "osd/OpQueueItem.h"
+#include "osd/mClockOpClassSupport.h"
+
+
+namespace ceph {
+
+  using Request = OpQueueItem;
+  using Client = uint64_t;
+
+  // This class exists to bridge the ceph code, which treats the class
+  // as the client, and the queue, where the class is
+  // osd_op_type_t. So this adapter class will transform calls
+  // appropriately.
+  class mClockPoolQueue : public OpQueue<Request, Client> {
+
+    using osd_op_type_t = ceph::mclock::osd_op_type_t;
+
+    using InnerClient = std::pair<int64_t,osd_op_type_t>;
+
+    using queue_t = mClockQueue<Request, InnerClient>;
+
+    queue_t queue;
+
+    ceph::mclock::OpClassClientInfoMgr client_info_mgr;
+
+    OSDService *service;
+
+  public:
+
+    mClockPoolQueue(CephContext *cct);
+
+    void set_osd_service(OSDService *service);
+
+    const crimson::dmclock::ClientInfo* op_class_client_info_f(const InnerClient& client);
+
+    inline unsigned length() const override final {
```

Comment:
i don't think the virtual functions of this class can be inlined in our case.

Reply:
As you say, pqueue is the parent class OpQueue, and it can not be written inline because it is not determined at compile time which child class it will be.
Removed the inline of all virtual functions in mClockClientQueue, mClockPoolQueue, and mClockOpClassQueue.
In addition, in the case of class mClockQueue, virtual functions are inlined because they are determined at compile time.